sql/mysql_db/jwt_auth.go

```diff
@@ -0,0 +1,240 @@
+	}
+	logrus.Info(logString)
+	return true, nil
+}
+
+func getClaimFromKey(claims jwt.Claims, field string) string {
+	switch field {
+	case "id":
+		return claims.ID
+	case "iss":
+		return claims.Issuer
+	case "sub":
+		return claims.Subject
+	}
+	return ""
+}
+
+func getExpectedClaims(claims map[string]string) jwt.Expected {
+	var expectedClaims jwt.Expected
+	for cl, value := range claims {
+		switch cl {
+		case "iss":
+			expectedClaims.Issuer = value
+		case "sub":
+			expectedClaims.Subject = value
+		case "aud":
+			expectedClaims.Audience = []string{value}
+		}
+	}
+	return expectedClaims
+}
+
+func getMatchingJwksConfig(config []JwksConfig, name string) (*JwksConfig, error) {
+	for _, item := range config {
+		if item.Name == name {
+			return &item, nil
+		}
+	}
+	return nil, fmt.Errorf("ValidateJWT: Matching JWKS config not found")
+}
+
+func getJWKSFromFile(filepath string) (*jose.JSONWebKeySet, error) {
+
+	file, err := os.Open(filepath)
+	if err != nil || file == nil {
+		return nil, errFileNotFound
+	}
+	defer file.Close()
+
+	byteValue, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	return unmarshalAndConvertKeys(byteValue)
+}
+
+func getJWKSFromUrl(locationUrl string) (*jose.JSONWebKeySet, error) {
+	client := &http.Client{}
+
+	request, err := http.NewRequest("GET", locationUrl, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	} else if response.StatusCode/100 != 2 {
+		return nil, errors.New("FetchedJWKS: Non-2xx status code from JWKS fetch")
+	} else {
+		defer response.Body.Close()
+		contents, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		return unmarshalAndConvertKeys(contents)
+	}
+}
+
+type jsonKey struct {
+	ID  string `json:"id"`
+	Key string `json:"key_bytes"`
+}
+
+func unmarshalAndConvertKeys(byteValue []byte) (*jose.JSONWebKeySet, error) {
+	jsonKeys := []jsonKey{}
+	err := json.Unmarshal(byteValue, &jsonKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make([]jose.JSONWebKey, len(jsonKeys))
+	for i, jKey := range jsonKeys {
+		priv, err := decodePrivateKey(jKey.Key)
+		if err != nil {
+			return nil, err
+		}
+		pub := priv.Public()
+		keys[i] = jose.JSONWebKey{
+			KeyID: jKey.ID,
+			Key:   pub,
+		}
+	}
+
+	return &jose.JSONWebKeySet{Keys: keys}, nil
+}
+
+func decodePrivateKey(key string) (*rsa.PrivateKey, error) {
+	dec, err := base32.StdEncoding.DecodeString(key)
+	if err != nil {
+		return nil, err
+	}
+	privKey, err := x509.ParsePKCS8PrivateKey(dec)
+	if err != nil {
+		return nil, err
+	}
+	k, ok := privKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, errors.New("could not convert to rsa.PrivateKey")
+	}
+	return k, nil
+}
+
+func getJWKSFromSource(source string) (*jose.JSONWebKeySet, error) {
+	// Check if file exists, otherwise get from url
+	jwks, err := getJWKSFromFile(source)
```

Comment:
Implementation of all of this would ideally live in dolt. But for that implementation, need a caching implementation here; something like we do for JWT auth in dolthubapi.